pkg/cloud/cloud.go

```diff
@@ -713,6 +809,56 @@ func (c *cloud) getVolumeId(ctx context.Context, volumeId string) (*string, erro
 	}
 }
 
+// getUpdateResizeFilesystemAdministrativeAction retrieves the AdministrativeAction associated with a file system update with the
+// given target storage capacity, if one exists.
+func (c *cloud) getUpdateResizeFilesystemAdministrativeAction(ctx context.Context, fileSystemId string, resizeGiB int64) (*fsx.AdministrativeAction, error) {
+	fs, err := c.getFileSystem(ctx, fileSystemId)
+	if err != nil {
+		return nil, fmt.Errorf("DescribeFileSystems failed: %v", err)
+	}
+
+	if len(fs.AdministrativeActions) == 0 {
+		return nil, fmt.Errorf("there is no update on filesystem %s", fileSystemId)
+	}
+
+	// AdministrativeActions are sorted from newest to oldest
+	for _, action := range fs.AdministrativeActions {
+		if *action.AdministrativeActionType == fsx.AdministrativeActionTypeFileSystemUpdate &&
+			action.TargetFileSystemValues.StorageCapacity != nil &&
+			*action.TargetFileSystemValues.StorageCapacity == resizeGiB {
+			return action, nil
+		}
+	}
+
+	return nil, fmt.Errorf("there is no update with storage capacity of %d GiB on filesystem %s", resizeGiB, fileSystemId)
+}
+
+// getUpdateResizeVolumeAdministrativeAction retrieves the AdministrativeAction associated with a volume update with the
+// given target storage capacity, if one exists.
+func (c *cloud) getUpdateResizeVolumeAdministrativeAction(ctx context.Context, volumeId string, resizeGiB int64) (*fsx.AdministrativeAction, error) {
+	v, err := c.getVolume(ctx, volumeId)
+	if err != nil {
+		return nil, fmt.Errorf("DescribeVolumes failed: %v", err)
+	}
+
+	if len(v.AdministrativeActions) == 0 {
+		return nil, fmt.Errorf("there is no update on volume %s", volumeId)
+	}
+
+	// AdministrativeActions are sorted from newest to oldest
+	for _, action := range v.AdministrativeActions {
+		if *action.AdministrativeActionType == fsx.AdministrativeActionTypeVolumeUpdate &&
+			action.TargetVolumeValues.OpenZFSConfiguration.StorageCapacityQuotaGiB != nil &&
```

Comment:
Should be verify that the OpenZFSConfiguration is not null? If (somehow) the volume id is associated with an Ontap volume, this would throw an NPE.

Reply:
Good catch, fixed